src/device/ping-device.cpp

```diff
@@ -0,0 +1,94 @@
+#include <ping-device.h>
+
+#include <ping-message-common.h>
+#include <ping-message.h>
+#include <ping-time.h>
+
+bool PingDevice::initialize() {
+    return request(CommonId::PROTOCOL_VERSION)
+        && request(CommonId::DEVICE_INFORMATION);
+}
+
+ping_message* PingDevice::read() {
+    char b;
+    int result = _port.read(&b, 1);
+    if (result > 0) {
+        if (_parser.parseByte(static_cast<char>(b)) == PingParser::NEW_MESSAGE) {
+        return &_parser.rxMessage;
+        }
+    }
+    return nullptr;
+}
+
```

Comment:
this is a bad style: who deletes the pointer?
it should probably return a unique_ptr or a std::vector<std::byte> if what we are reading are byte streams.

Reply:
note that this is a migration of an existing implementation to reduce maintainence overhead.

I take your point on pointer management and poor style.

The pointer management here is documented in doxygen in the header file.

I am intentionally avoiding standard library as not all of our supported platforms support it (arduino).